File: julia/pseudoinverse.jl
@doc """
This algorithm solves the Thikonov regularized least squares Problem 
argminₓ(‖Ax-b‖² + λ‖b‖²) using the singular value decomposition of A.

# Arguments

* `U, Σ, V`: Singular value decomposition of A
* `b::Vector`: Measurement vector b
* `lambd::Float64`: The regularization parameter, relative to the matrix trace
* `enforceReal::Bool`: Enable projection of solution on real plane during iteration
* `enforcePositive::Bool`: Enable projection of solution onto positive halfplane during iteration
""" ->
function pseudoinverse{T}(U::Matrix, Σ::Vector, V::Matrix, b::Vector{T}, lambd, enforceReal, enforcePositive)
	# perform regularization
	D = zeros(Σ)
	for i=1:length(Σ)
		σi = Σ[i]
		D[i] = σi/(σi^2+lambd^2)
	end

	# calculate pseudoinverse
	tmp = BLAS.gemv('C', one(T), U, b)
	tmp .*=  D
	c = BLAS.gemv('N', one(T), V, tmp)

	# apply constraints
	if enforceReal && eltype(c) <: Complex
		c = complex.(real.(c),0)
	end
	if enforcePositive
		c[real(c) .< 0] = 0
	end
	
	return c
end

File: python/pseudoinverse.py
import numpy as np
import scipy as sp

"""
This algorithm solves the Thikonov regularized least squares problem 
using the singular value decomposition of A.

# Arguments

* `U, Sigm, V`: Singular value decomposition of A
* `b::Vector`: Measurement vector b
* `lambd::Float64`: The regularization parameter, relative to the matrix trace
* `enforceReal::Bool`: Enable projection of solution on real plane during iteration
* `enforcePositive::Bool`: Enable projection of solution onto positive halfplane during iteration
"""
def pseudoinverse(U, Sigm, V, b, lambd, enforceReal, enforcePositive):
  # perform regularization
  D = np.zeros(np.size(Sigm))
  for i in range(np.size(Sigm)):
    sigmi = Sigm[i]
    D[i] = sigmi/(sigmi**2+lambd**2)

  # calculate pseudoinverse
  tmp = np.dot(U.conjugate().transpose(),b[:]) #  conjugate transpose
  tmp =  tmp*D
  c = np.dot(V.conjugate().transpose(),tmp) # not transposed

  if enforceReal and np.iscomplexobj(c):
    c.imag = 0
  if enforcePositive:
    c = c * (c.real > 0)

  return c
